packages/hashtower/src/hashtower-hash-chain-proof-builder.ts

```diff
@@ -0,0 +1,99 @@
+import checkParameter from "./checkParameter"
+import { HashFunction, HashTowerHashChainProof } from "./types"
+
+const pad = (arr: any, len: number, val: any) => arr.concat(Array(len - arr.length).fill(val))
+const pad0 = (arr: any, len: number) => pad(arr, len, BigInt(0))
+const pad00 = (arr2D: any, h: number, w: number) => pad(arr2D, h, []).map((a: any) => pad0(a, w))
+
+/**
+ * HashTowerHashChainProofBuilder is a TypeScript implementation of HashTower to generate proofs of membership.
+ * @param H Height of tower of the proving circuit. It can be less than the H in the contract.
+ * @param W Width of tower.
+ * @param hash A hash function which supports 2 input values.
+ */
+export default function HashTowerHashChainProofBuilder(H: number, W: number, hash: HashFunction) {
```

Comment:
What about a JS class here?

Reply:
I can't remember the specific reason but using separate functions made the algorithm faster. I think there were no differences between function and class tho. That code should be compiled into JS functions anyway in theory

Reply:
I did some benchmarking, and there was no significant difference.

I pushed a new TS-class commit.